Swagger documentation gsoc2019

kangbreder opened the pull request:
## Description
For GSoC, I have improved the Swagger document with the following

- Reflected all API descriptions on swagger UI

- Formatted API descriptions

- Added swagger docs for new APIs

- Added swagger-gradle-plugin for automatic save of json during build.

- Leveraged swagger codegen for generation of language-specific libraries using gradle-swagger-generator plugin.

Link to JIRA TICKET: https://jira.apache.org/jira/browse/FINERACT-733

- [COMPLETED] Complete Swagger documentation on all API’s of Fineract

- [COMPLETED] Description inside of swagger docs

- [IN PROGRESS] Try it out button for API’s from swagger-UI

- [IN PROGRESS] Automatically generate the swagger-docs -> building with Fineract

- [NOT COMPLETED] Navigation Bar on swagger-UI(For easy access of documentation of API's)

- [NOT COMPLETED] Migrating the complete overview section in swagger-docs/swagger-docs

- [COMPLETED] Swagger guide for new Api’s or how to add a swagger doc to any Fineract API

- [IN PROGRESS] QA round to check if all the API’s are reflecting or not(Correct)


## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Commit message starts with the issue number from https://issues.apache.org/jira/projects/FINERACT/. Ex: FINERACT-646 Pockets API.

- [x] Coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions have been followed.

- [x] API documentation at https://github.com/apache/fineract/blob/develop/api-docs/apiLive.htm has been updated with details of any API changes.

- [ ] Integration tests have been created/updated for verifying the changes made.

- [x] All Integrations tests are passing with the new commits.

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the list.)

Our guidelines for code reviews is at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide

kangbreder committed:
Reflected all API descriptions on swagger UI.

Formatted API descriptions.

Added swagger docs for new APIs.

Added swagger-gradle-plugin for automatic save of json during build.

Leveraged swagger codegen for generation of language-specific libraries using gradle-swagger-generator-plugin

kangbreder committed:
Reflected all API descriptions on swagger UI.

Formatted API descriptions.

Added swagger docs for new APIs.

Added swagger-gradle-plugin for automatic save of json during build.

Leveraged swagger codegen for generation of language-specific libraries using gradle-swagger-generator plugin.

Sanyam96 commented:
Adding Authentication and common params for swagger documentation to each and every endpoint(API's ) is unnecessary and is not a good coding standard.
It will increase redundant code for auth to every API's, which can be done by sharing a common resource to all the API's

In the review of removing Auth params from every API in your commit: 
https://github.com/kangbreder/fineract/commit/7418ac403e4e145c86dd826086a2799dd5c4bf7c#commitcomment-34794554

Here are some reference links which you can refer to remove those params!
https://stackoverflow.com/questions/47627469/swagger2-too-many-the-same-annotations
https://stackoverflow.com/questions/39853217/swagger-how-to-combine-shared-information-from-different-apis

kangbreder commented:
Okay. I will consider that and let you know.

On Mon, Aug 26, 2019 at 6:44 AM Sanyam Goel <notifications@github.com>
wrote:

> Adding Authentication and common params for swagger documentation to each
> and every endpoint(API's ) is unnecessary and is not a good coding standard.
> It will increase redundant code for auth to every API's, which can be done
> by sharing a common resource to all the API's
>
> In the review of removing Auth params from every API in your commit:
> kangbreder@7418ac4#commitcomment-34794554
> <https://github.com/kangbreder/fineract/commit/7418ac403e4e145c86dd826086a2799dd5c4bf7c#commitcomment-34794554>
>
> Here are some reference links which you can refer to remove those params!
>
> https://stackoverflow.com/questions/47627469/swagger2-too-many-the-same-annotations
>
> https://stackoverflow.com/questions/39853217/swagger-how-to-combine-shared-information-from-different-apis
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/629?email_source=notifications&email_token=AJOPQRCXQ3APLX4MAREQZSDQGNUT5A5CNFSM4IPIBQPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5DKNRQ#issuecomment-524723910>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJOPQRBVQEUR6IRH4SBEE7LQGNUT5ANCNFSM4IPIBQPA>
> .
>

Sanyam96 commented:
I don't think, we should require the below files in the codebase
catalina.base_IS_UNDEFINED/logs/fineract-platform.log
.response.json.swp

if we need anyhow,
please try to add some documentation regarding those files

Sanyam96 commented:
I don't think, we should require the below files in the codebase
```catalina.base_IS_UNDEFINED/logs/fineract-platform.log```
```.response.json.swp```

if we need anyhow,
please try to add some documentation regarding those files

Sanyam96 commented on fineract-provider/src/main/java/org/apache/fineract/accounting/accrual/api/AccrualAccountingApiResource.java:
```diff
@@ -38,7 +38,11 @@
+@SwaggerDefinition(tags = {
+        @Tag(name = "Periodic Accrual Accounting", description = "Periodic Accrual is to accrue the loan income till the specific date or till batch job scheduled time.\n\n")
+})
+
```
I don't think, we need to add an extra blank line here

Sanyam96 replied:
Still seeing blank lines!

Sanyam96 commented on fineract-provider/src/main/java/org/apache/fineract/accounting/closure/api/GLClosuresApiResource.java:
```diff
@@ -55,7 +55,11 @@
+@SwaggerDefinition(tags = {
+        @Tag(name = "Accounting Closure", description = "An accounting closure indicates that no more journal entries may be logged (or reversed) in the system, either manually or via the portfolio with an entry date prior to the defined closure date\\n\" + \"\\n\" + \"Field Descriptions\\n\" + \"closingDate\\n\" + \"The date for which the accounting closure is defined\\n\" + \"officeId\\n\" + \" ...
+})
+
```
I don't think, we need to add an extra blank line here

Sanyam96 commented on fineract-provider/src/main/java/org/apache/fineract/accounting/financialactivityaccount/api/FinancialActivityAccountsApiResource.java:
```diff
@@ -51,7 +51,11 @@
+@SwaggerDefinition(tags = {
+        @Tag(name = "Mapping Financial Activities to Accounts", description = "Organization Level Financial Activities like Asset and Liability Transfer can be mapped to GL Account. Integrated accounting takes these accounts into consideration when an Account transfer is made between a savings to loan/savings account and vice-versa\\n\" + \"\\n\" +\"\\nField Descriptions\\n\" +\"financialActivity ...
+})
+
```
I don't think, we need to add an extra blank line here

Sanyam96 commented on fineract-provider/src/main/java/org/apache/fineract/organisation/workingdays/api/WorkingDaysApiResource.java:
```diff
@@ -45,7 +45,11 @@
-@Api(value = "Working days", description = "The days of the week that are workdays.\n" + "\n" + "Rescheduling of repayments when it falls on a non-working is turned on /off by enable/disable reschedule-future-repayments parameter in Global configurations.\n" + "\n" + "Allow transactions on non-working days is configurable by enabling/disbaling the allow-transactions-on-non_workingday parameter in ...
+@Api(tags = {"Working days"})
+@SwaggerDefinition(tags = {
+        @Tag(name = "Working days", description = "The days of the week that are workdays.\\n\" + \"\\n\" + \"Rescheduling of repayments when it falls on a non-working is turned on /off by enable/disable reschedule-future-repayments parameter in Global configurations.\\n\" + \"\\n\" + \"Allow transactions on non-working days is configurable by enabling/disbaling the allow-transactions-on-non_work ...
```
I am able to ```\n``` in swagger-ui,
try to remove all the unnecessary letters in the description

You can cross - verify for all the API's from swagger-UI

Sanyam96 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/floatingrates/api/FloatingRatesApiResource.java:
```diff
@@ -45,7 +45,11 @@
-@Api(value = "Floating Rates", description = "It lets you create, list, retrieve and upload the floating rates")
+@Api(tags = {"Floating Rates"})
+@SwaggerDefinition(tags = {
+		@Tag(name = "Floating Rates", description = "It lets you create, list, retrieve and upload the floating rates")
```
Indentation of ```@Tag``` annotation is not correct, please correct!

kangbreder replied:
For swagger's API documentation, here is the way to add description 
@Api(tags = {"Swagger Resource"})
@SwaggerDefinition(tags = {
    @Tag(name = "Swagger Resource", description = "Write description here")
})
public class ... {
}

That is what i got from swagger github issues.
does that mean i will have to unindent? I have not checked if that will work but right now, that description properly appears on the UI

Sanyam96 replied:
I'm seeing double ```tab-blank``` spaces in annotation 
```@Tag(name = "Floating Rates", description = "It lets you create, list, retrieve and upload the floating rates")```

vorburger replied:
@Sanyam96 I would not worry too much about this kind of thing (whitespace and indent), we're about to have a general solution for this, please watch [FINERACT-821](https://issues.apache.org/jira/browse/FINERACT-821). (Help from you on that one would be very welcome BTW!)

Sanyam96 commented:
@kangbreder Please verify all the description of each and every API from swagger-UI.
and try to remove all the unnecessary letters/alphabets/characters from the code base

Sanyam96 commented:
@kangbreder I am still seeing blank lines on API's 

<img width="603" alt="Screenshot 2019-09-01 at 8 56 22 PM" src="https://user-images.githubusercontent.com/11756370/64078663-10124080-ccfb-11e9-989b-9f5e599437fe.png">

You have also not given your explanation on https://github.com/apache/fineract/pull/629#issuecomment-524767338

kangbreder commented:
> I don't think, we should require the below files in the codebase
> `catalina.base_IS_UNDEFINED/logs/fineract-platform.log`
> `.response.json.swp`
> 
> if we need anyhow,
> please try to add some documentation regarding those files

I think these files were auto generated during build some time back.

vorburger commented:
kangbreder@ I just merged your two earlier smaller PRs. Will you rebase this one and resolve conflicts?

Sanyam96 commented:
@vorburger 
I think this PR is not completed yet and a lot of work is still needed to be done.

and this PR also needs to be tested before merging into the codebase.



@kangbreder can you please share the status of this PR?

Sanyam96 commented:
@kangbreder 
I can still able to view ```.response.json.swp``` file and ```catalina.base_IS_UNDEFINED/logs/fineract-platform.log``` file

Can you please remove these two files, as I think these are not relevant

Sanyam96 reviewed (changes requested):
@kangbreder
I can still able to view .response.json.swp file and catalina.base_IS_UNDEFINED/logs/fineract-platform.log file

Can you please remove these two files, as I think these are not relevant

kangbreder commented:
> kangbreder@ I just merged your two earlier smaller PRs. Will you rebase this one and resolve conflicts?

@vorburger I am currently resolving the conflicts. I will send an update soon. This PR is not ready for a merge yet. I am currently still working on this.

vorburger commented:
@kangbreder @Sanyam96 how are you progressing with resolving the conflicts and making this ready to merge? If you were able to have complete Swagger documentation finished ASAP so that it could still make it into our 1.4.0 release that would be great!

kangbreder commented:
I have been resolving the conflicts and they keep happening. I keep
resolving conflicts I had already resolved.

On Thu, Dec 12, 2019, 08:15 Michael Vorburger ⛑️ <notifications@github.com>
wrote:

> @kangbreder <https://github.com/kangbreder> @Sanyam96
> <https://github.com/Sanyam96> how are you progressing with resolving the
> conflicts and making this ready to merge? If you were able to have complete
> Swagger documentation finished ASAP so that it could still make it into our
> 1.4.0 release that would be great!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/629?email_source=notifications&email_token=AJOPQRENPMDTEV4O3LFLNYDQYHQJRA5CNFSM4IPIBQPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGVWR7A#issuecomment-564881660>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJOPQRANVFNZRTTKZJNOLITQYHQJRANCNFSM4IPIBQPA>
> .
>

vorburger commented:
> I have been resolving the conflicts and they keep happening. 
> I keep resolving conflicts I had already resolved.

Would you like assistance and have one of us help with resolving these conflicts (in new PR) ?  If so, perhaps this is something your GSoC mentor could support you with? I'm not sure who that was, was it @Sanyam96 or someone else? Otherwise, an active commiter such as e.g. @awasum or myself or @mohitsinha could perhaps take this over? I think it would be great if we could wrap up this PR by let's say end of next week latest.

kangbreder commented:
Okay. I will connect with Sanyam this weekend.

On Fri, Dec 13, 2019, 10:31 Michael Vorburger ⛑️ <notifications@github.com>
wrote:

> I have been resolving the conflicts and they keep happening.
> I keep resolving conflicts I had already resolved.
>
> Would you like assistance and have one of us help with resolving these
> conflicts (in new PR) ? If so, perhaps this is something your GSoC mentor
> could support you with? I'm not sure who that was, was it @Sanyam96
> <https://github.com/Sanyam96> or someone else? Otherwise, an active
> commiter such as e.g. @awasum <https://github.com/awasum> or myself or
> @mohitsinha <https://github.com/mohitsinha> could perhaps take this over?
> I think it would be great if we could wrap up this PR by let's say end of
> next week latest.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/629?email_source=notifications&email_token=AJOPQRBQTONQIYHHPCX42TLQYNI7TA5CNFSM4IPIBQPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGZNVWA#issuecomment-565369560>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJOPQRGHQ4ZRCIFUTMXT3FDQYNI7TANCNFSM4IPIBQPA>
> .
>

Sanyam96 commented:
@kangbreder I would recommend completing the project or making a project to a stable state then resolving the conflicts at the final state.
so that we can release the done part of the docs in the next release

kangbreder commented:
I have resolved the conflicts on the PR but having this build failure.
Here are the logs https://pastebin.com/DKzcGgZE
Please can i get some help with this?

On Mon, Dec 16, 2019 at 11:32 AM Sanyam Goel <notifications@github.com>
wrote:

> @kangbreder <https://github.com/kangbreder> I would recommend completing
> the project or making a project to a stable state then resolving the
> conflicts at the final state.
> so that we can release the done part of the docs in the next release
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/629?email_source=notifications&email_token=AJOPQRC5L7J63EDEX62EZR3QY5KKZA5CNFSM4IPIBQPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG6IBFY#issuecomment-566001815>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJOPQRDB5BGJ7XQOL3KAJADQY5KKZANCNFSM4IPIBQPA>
> .
>

awasum commented:
Push the new changes to yours branch and they will show here. then someone might be able to help.

kangbreder commented:
Okay. I have done that

On Thu, Dec 19, 2019 at 12:02 PM Awasum Yannick <notifications@github.com>
wrote:

> Push the new changes to yours branch and they will show here. then someone
> might be able to help.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/629?email_source=notifications&email_token=AJOPQRHZ5DRPJC4IV6Q53M3QZNIEBA5CNFSM4IPIBQPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHJH42Q#issuecomment-567443050>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJOPQRAI44DJARAG7PXVC63QZNIEBANCNFSM4IPIBQPA>
> .
>

kangbreder commented:
I have resolved the conflicts on the PR but having this build failure.
Here are the logs https://pastebin.com/DKzcGgZE
Please can i get some help with this?

vorburger commented:
> having this build failure. Please can i get some help with this?

This is because this PR, as is, also modifies *.gradle build script files.. and many more, check out the _Files change_ and _Commits_ tabs on this PR, and you'll see what I mean, it looks like this would revert unrelated previous changes - which I'm sure isn't what you are proposing here, but something in your local git operations has gone wrong. @Sanyam96 are you able to help @kangbreder to get this to a clean state? Please feel free to shout if you need me or @awasum to assist.

vorburger commented:
@kangbreder the solution you are going for here seems to be to manually clean up the build script with additional commits? What you really want to do instead is to not have that many commits (155!), and make this PR only have your changes... have you tried to simply do a `git rebase` instead?

kangbreder commented:
Yea i did a rebase before i got the build failure

On Sat, Jan 4, 2020 at 1:19 PM Michael Vorburger ⛑️ <
notifications@github.com> wrote:

> @kangbreder <https://github.com/kangbreder> the solution you are going
> for here seems to be to manually clean up the build script with additional
> commits? What you really want to do instead is to not have that many
> commits (155!), and make this PR only have your changes... have you tried
> to simply do a git rebase instead?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/629?email_source=notifications&email_token=AJOPQREF2HF3ZEEZFO4DFB3Q4B5GXA5CNFSM4IPIBQPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEICW3BI#issuecomment-570781061>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJOPQRCKHTTGJRGZNOHQBF3Q4B5GXANCNFSM4IPIBQPA>
> .
>

kangbreder commented:
> @kangbreder the solution you are going for here seems to be to manually clean up the build script with additional commits? What you really want to do instead is to not have that many commits (155!), and make this PR only have your changes... have you tried to simply do a `git rebase` instead?

@vorburger Yea i had done a rebase but after that, i got the build failure. So, i manually integrated the changes from develop branch.

Sanyam96 commented:
@kangbreder Can you please update the status of the project regarding the Jira ticket?
https://jira.apache.org/jira/browse/FINERACT-733

Sanyam96 commented:
@kangbreder I am also able to see lot of files, which are not related to your changes.
Can you please remove those and make this PR more relevant to the Swagger docs?

Also can you please squash your commits into a single commit?

Sanyam96 commented on .gitignore:
```diff
@@ -9,4 +9,5 @@ build
-gradle
\ No newline at end of file
+gradle
+catalina.base_IS_UNDEFINED/
```
Please remove this

Sanyam96 reviewed (changes requested):
@kangbreder Please remove all the files, which are not related to swagger
It seems that some commits from the master are merged into this PR, Maybe you can use
```git rebase``` command to remove all the commits, which are not associated with swagger-docs.

And kindly update the status of the swagger docs project.

kangbreder committed:
removed unnecessary files

vorburger commented:
@kangbreder this PR is now looking much much better than ever before - congratulations!

There are just 2 conflicts, due to the recent merge of #678 - could you manually resolve those? Once those conflicts are resolved and when this passes the builds, then this looks good to me, and I'm happy to merge it ASAP - let's still get this into release 1.4.0! @Sanyam96 @awasum  does this have LGTM from you guys as well, as-is?

PS: One thing I would love to understand better, and I'm sure others would like to as well, is how this will actually be used? I'm sure to you this is "obvious", but to all those of us who have not been involved in the details this GSOC project, would it be possible to have some "doc" for this? It doesn't matter if initially it's given here or in [FINERACT-733](https://jira.apache.org/jira/browse/FINERACT-733) or may be somewhere on the Apache Confluence Wiki. Just to understand, once we merge this, what does this actually "do" - you know what I mean? I gather these two new Gradle plugins you are adding ([io.swagger.core.v3.swagger-gradle-plugin](https://github.com/swagger-api/swagger-core/tree/master/modules/swagger-gradle-plugin) and [org.hidetake.swagger.generator](https://github.com/int128/gradle-swagger-generator-plugin)) do... what - just generate the src/main/resources/swagger-ui/response.json file? Can I ask why we need 2 plugins to generate 1 file? :smiling_imp: Or do these to more

I can see that file already exists now, but I'm not clear where that's coming from, and how it was initially built.. does someone else know? @vishwasbabu or @Sanyam96 perhaps? Is that this PR contributes, in essence, really the automatic build of that file?

If that new src/main/resources/swagger-ui/response.json file is (now) a generated artifact, then why is it in Git, and generated under src/ instead of just being put in the right place under build/ so that it ends up in the WAR? Perhaps that could be contributed as a follow-up once we have merged this?

kangbreder committed:
Merge branch 'develop' into swagger-documentation-gsoc2019

kangbreder commented:
@vorburger Thank you. I have resolved the conflicts. 

The response.json file is the OpenAPI specification file which is used to generate the swagger-ui. 
The already existing response.json file in src/main/resources/swagger-ui/ is the spec generated with the request https://localhost:8443/fineract-provider/api/v1/swagger.json. Every time a modification is done on the swagger docs or a new swagger doc is added, this spec needs to be generated again at runtime and the response.json file as well updated so that the modifications can be seen on the UI. 

The io.swagger.core.v3.swagger-gradle-plugin is a plugin i had added to automatically generate the specs during build time and not at run time. This was so that every time one builds on the project, the response.json updates. But this did not work as expected. I had removed the plugin from the build script. I am wondering how its still in there. probably while i have rebasing and resolving conflicts, that came back in the script. I will remove the plugin. 

The other plugin org.hidetake.swagger.generator is used to generate code from the openAPI spec file using swagger-codegen. So this does not generate response.json. It rather uses response.json to generate client code. You

@vorburger i think you have a nice idea. Is it possible to get the file and the swagger-ui in the war? 
That is actually what is needed to be done. Presently the file is manually generated and updated each time.

Sanyam96 commented:
@kangbreder thank you for making the PR more stable.

I'll test your changes very soon so that we can merge and release your changes in the next cycle release cycle.

@vorburger 
I also agree with you regarding usability.
I think we should document somewhere, maybe at a centric place and also provide the link in the readme so that other devs can refer the swagger-docs and how to use swagger docs in the future.


I am also not sure, about the plugins, do we require it or not, right now.
I will review this PR, and provide my feedback and required changes to this PR.

vorburger commented:
> @vorburger Thank you. I have resolved the conflicts.

following the merge of #682, there now is a new conflict again already... will you resolve, again?

I think to avoid that happening again and again, we should try to get a minimal version of this PR merged ASAP, and then incrementally iterate on improving it in separate follow-up PRs.

> The io.swagger.core.v3.swagger-gradle-plugin is a plugin i had added to automatically generate the specs during build time and not at run time. This was so that every time one builds on the project, the response.json updates. But this did not work as expected. I had removed the plugin from the build script. I am wondering how its still in there. probably while i have rebasing and resolving conflicts, that came back in the script. I will remove the plugin.

OK, cool. From my side, we could perhaps merge this PR after you've done that (but @Sanyam96 more than me should fully review and LGTM this).

> @vorburger i think you have a nice idea. Is it possible to get the file and the swagger-ui in the war?

Yeah, I think it should be. @Sanyam96 can probably work with you towards that. Again, I would suggest that we get this PR to a state where it can be merged even if it can still be improved in follow-up you can raise in separate new PRs.

kangbreder committed:
Merge branch 'develop' into swagger-documentation-gsoc2019

awasum commented:
@Sanyam96 , @vorburger , Seems conflicts are now fixed by @kangbreder  and Travis is passing. Merge this when yo feel comfortable... instead create more issues and enhancement tickets related to this feature on Jira for @kangbreder  to pick up and fix gradually.

vorburger reviewed (approved):
@Sanyam96 does this have LGTM from you? From my side, I'm fully aligned with @awasum - let us finally get this merged ASAP, and then gradually improve upon it (I do understand the work expected here is not fully done yet).

I'm going to try to have this one automagically rebased using #691, curious if this will work (1st time):

/rebase

vorburger commented on fineract-provider/build.gradle:
```diff
@@ -59,6 +60,8 @@ apply plugin: 'eclipse'
 apply plugin: 'idea'
 apply plugin: 'com.bmuschko.tomcat'
 apply plugin: 'project-report'
+apply plugin: 'java'
```
@kangbreder do you happen to remember why you had to add `apply plugin: 'java'` here? Does it work without it as well? I guess it's a bit strange to have a project use BOTH `apply plugin: 'java-library'` AND `apply plugin: 'java'` - it's either or the other.

Sanyam96 replied:
I also think this is not relevant to this swagger changes
@kangbreder can you please verify this.

kangbreder replied:
> @kangbreder do you happen to remember why you had to add `apply plugin: 'java'` here? Does it work without it as well? I guess it's a bit strange to have a project use BOTH `apply plugin: 'java-library'` AND `apply plugin: 'java'` - it's either or the other.

@vorburger I have not checking if it works without the plugin. My Ubuntu 18 has failed to boot and i have been trying to correct the problem. Because of that, I have not been able to check that yet

vorburger commented:
/rebase

vorburger commented:
FYI: /rebase does not actually seem to work just yet, see [FINERACT-829](https://issues.apache.org/jira/browse/FINERACT-829) / [INFRA-19719](https://issues.apache.org/jira/browse/INFRA-19719). We can manually just manually rebase it, once we have the LGTM from @Sanyam96 on this one. (Or we just could merge it with a merge commit; not my favourite, but also fine.)

Sanyam96 commented on fineract-provider/src/main/java/org/apache/fineract/mix/api/MixTaxonomyMappingApiResource.java:
```diff
@@ -31,7 +31,9 @@
 import javax.ws.rs.core.UriInfo;
 
+
 import io.swagger.annotations.Api;
```
remove this import
I think it should be covered in line 36

Sanyam96 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/search/api/SearchApiResource.java:
```diff
@@ -85,8 +93,10 @@ public String retrieveAdHocSearchQueryTemplate(@Context final UriInfo uriInfo) {
-            @QueryParam("resource") final String resource ,@DefaultValue("false") @QueryParam("exactMatch")  Boolean exactMatch) {
+    @ApiOperation(value = "Search Resources", notes = "Example Requests:\n" + "\n" + "search?query=000000001\n" + "\n" + "\n" + "search?query=Petra&resource=clients,groups\n" + "\n" + "\n" + "search?query=Petra&resource=clients,groups&exactMatch=true")
+    @ApiResponses({@ApiResponse(code = 200, message = "", response = SearchApiResourceSwagger.GetSearchResponse.class)})
+    public String searchData(@Context final UriInfo uriInfo, @QueryParam("query")  @ApiParam(value = "query")final String query,
```
Please add space between
```suggestion
    public String searchData(@Context final UriInfo uriInfo, @QueryParam("query")  @ApiParam(value = "query") final String query,
```

Sanyam96 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/self/shareaccounts/api/SelfShareAccountsApiResource.java:
```diff
@@ -139,7 +152,11 @@ public String createAccount(final String apiRequestBodyAsJson) {
 	@Consumes({ MediaType.APPLICATION_JSON })
 	@Produces({ MediaType.APPLICATION_JSON })
-	public String retrieveShareAccount(@PathParam("accountId") final Long accountId, @Context final UriInfo uriInfo) {
+	/*
+    @ApiOperation(value = "Retrieve a share application/account", httpMethod = "GET", notes = "\n" +
+            "\n" + "\n" + "Example Requests:\n" + "\n" + "self/shareaccounts/12\n")
+    @ApiResponses({@ApiResponse(code = 200, message = "OK", responseContainer = "List", response  SelfShareAccountsApiResourceSwagger.GetShareAccountResponse.class)})*/
```
Why these annotations is commented??

vorburger replied:
@Sanyam96 FYI I've fixed this one in my swagger-squash branch which I'll merge ASAP as much as possible. @kangbreder the problem is that SelfShareAccountsApiResourceSwagger.GetShareAccountResponse seems to be missing- would you like to raise a small new PR just adding that and fixing this?

Sanyam96 commented on fineract-provider/build.gradle:
```diff
@@ -42,6 +42,7 @@ buildscript {
      classpath "org.apache.openjpa:openjpa:$openJPAVersion"
      classpath 'com.radcortez.gradle:openjpa-gradle-plugin:3.1.0'
      classpath 'gradle.plugin.org.nosphere.apache:creadur-rat-gradle:0.2.2'
+      classpath "gradle.plugin.org.hidetake:gradle-swagger-generator-plugin:2.18.1"
```
I think this plugin would not help to generate the swagger.json automatically during project build
please remove this.

Sanyam96 commented on fineract-provider/build.gradle:
```diff
@@ -529,9 +533,26 @@ bootWar {
 }
 
+//This is for code and swagger UI generation
+dependencies {
+    swaggerCodegen 'io.swagger:swagger-codegen-cli:2.4.2'             // Swagger Codegen V2
+    swaggerCodegen 'io.swagger.codegen.v3:swagger-codegen-cli:3.0.5'  // or Swagger Codegen V3
+    swaggerCodegen 'org.openapitools:openapi-generator-cli:3.3.4'     // or OpenAPI Generator
+    swaggerUI 'org.webjars:swagger-ui:3.10.0'
+}
+
+swaggerSources {
+    fineract {
+        inputFile = file('src/main/resources/swagger-ui/response.json')
+        code {
+            language = 'java'
+        }
+    }
+}
+
```
@kangbreder  Can you please clarify about these changes?
I am not sure how this code will help us?.

kangbreder replied:
> @kangbreder Can you please clarify about these changes?
> I am not sure how this code will help us?.

@Sanyam96  have you tried using this to see how it works? Are you sure you really want me to remove all changes made on the build.script. Are you sure about that?

kangbreder replied:
> @kangbreder Can you please clarify about these changes?
> I am not sure how this code will help us?.

@Sanyam96 I think i had explained the usage of this in my gist https://gist.github.com/kangbreder/034f47e2e8015cee10b48b7c5f1b8df1, generating client libraries and language-specific SDKs in languages such as java and angular from swagger specification file. This was part of the goals of the project listed in https://mifosforge.jira.com/wiki/spaces/RES/pages/812810251/Google+Summer+of+Code+2019+Ideas. 

I clarified the usage of this this to @vorburger the other day. I had removed the other plugin which was added for generation of spec file during build. Since that did not work, it has been removed it from the PR. 

I suggest you verify that again.
 My ubuntu 18 will not boot.
 I have been trying to solve the problem. Because of that, i have not been able to do the modifications.

Sanyam96 replied:
Thanks for the clarification @kangbreder .

Please do other requested changes other than build.gradle.

Sanyam96 commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/reportmailingjob/api/ReportMailingJobRunHistoryApiResource.java:
```diff
@@ -43,8 +43,10 @@
 
 @Path("/" + ReportMailingJobConstants.REPORT_MAILING_JOB_RUN_HISTORY_RESOURCE_NAME)
 @Component
-@Scope("singleton")
```
Why it is removed?

Sanyam96 commented:
@kangbreder Please do the following changes in the PR.

1. Please remove all the changes from the build.gradle file.
2. Please make the changes mentioned above in my review.
3. Please squash all the commits into a single commit.

@vorburger @awasum 
I think apart from the above-mentioned changes Everything else Looks Good To Me.
After Kang made all the changes, we can work to merge this PR.

vorburger commented:
@kangbreder because I'd like to help you avoid running into painful whitespace related merge issues due to #688 which I'm currently working on and would like to wrap up today, here's what I'm going to do about this PR:

I'll local pull this, and make the changes which @Sanyam96 suggests, and merge it ASAP. (This will allow me to then rebase #688 and do the overall whitespace clean up, which will unblock #689 which I'd like to get in rather sooner than later.)

vorburger commented:
Closing this PR as having been replaced with #695 (which is basically this, with Authorship by Kang retained in the git history of course), just squashed all into 1 single commit, followed up by a commit from me which implements the requested changes.

[FINERACT-835](https://jira.apache.org/jira/browse/FINERACT-835) now includes a number of more specific new sub-tasks with proposed follow-ups.
